llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -3668,6 +3668,51 @@ static bool hasAffectedValue(Value *V, SmallPtrSetImpl<Value *> &Affected,
   return false;
 }
 
+static Value *foldSelectAddConstant(SelectInst &SI,
+                                    InstCombiner::BuilderTy &Builder) {
+  Value *Cmp;
+  Instruction *FAdd;
+  ConstantFP *C;
+
+  // select((fcmp OGT/OLT, X, 0), (fadd X, C), C) => fadd((select (fcmp OGT/OLT,
+  // X, 0), X, 0), C)
+
+  // This transformation enables the possibility of transforming fcmp + sel into
+  // a fmax/fmin.
+
+  // OneUse check for `Cmp` is necessary because it makes sure that other
+  // InstCombine folds don't undo this transformation and cause an infinite
+  // loop.
+  if (match(&SI, m_Select(m_OneUse(m_Value(Cmp)), m_OneUse(m_Instruction(FAdd)),
+                          m_ConstantFP(C))) ||
+      match(&SI, m_Select(m_OneUse(m_Value(Cmp)), m_ConstantFP(C),
+                          m_OneUse(m_Instruction(FAdd))))) {
+    Value *X;
+    CmpInst::Predicate Pred;
+    if (!match(Cmp, m_FCmp(Pred, m_Value(X), m_AnyZeroFP())))
+      return nullptr;
+
+    if (Pred != CmpInst::FCMP_OGT && Pred != CmpInst::FCMP_OLT)
+      return nullptr;
+
+    if (!match(FAdd, m_FAdd(m_Specific(X), m_Specific(C))))
+      return nullptr;
+
+    FastMathFlags FMF = FAdd->getFastMathFlags();
+    FMF |= SI.getFastMathFlags();
+
+    Value *NewSelect = Builder.CreateSelect(
+        Cmp, X, ConstantFP::getZero(C->getType()), SI.getName() + ".new", &SI);
```

Comment:
The identity value for fadd is -0, not +0. This transform as written is incorrect for -0 (but I believe is correct if you just emit -0 here)

Reply:
We will need to emit the `fcmp` instruction with `-0` as well otherwise two things will happen:
1. The IC [transformation](https://github.com/llvm/llvm-project/blob/d6ad55167094b7b23e71ec6eabfe942e7a485b41/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L1685) will not break at the [condition ](https://github.com/llvm/llvm-project/blob/d6ad55167094b7b23e71ec6eabfe942e7a485b41/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L1707) and cause an infinite loop.
2. The `fcmp + sel` combination will not lower into `fmax/fmin`. https://godbolt.org/z/7r3Mnzq41

I believe it would be safe, but I wanted to verify if there's anything I might be overlooking.

Reply:
It's not useful to emit fcmp with a -0. Fcmp with 0 or -0 are exactly the same operation

Reply:
I would emit the -0, and only emit the 0 with an NSZ flag